Make pull request remediations extendable, add action pinning remediation

jhrozek opened the pull request:
Our PR remediations only allowed replacing content. In order to make the
PR remediations more usable, let's add a small extendable library. As
the first step, the existing content is retained, just moved to that
library.

Additionally, a new method is added that calls the frizbee library to
replace tags in workflows with SHA pinning. To use this new method,
you'll have to use a patched `actions_check_pinned_tags` rule_type.

Fixes: #1774
Fixes: #1807

jhrozek commented on proto/minder/v1/minder.proto:
```diff
@@ -1282,17 +1282,22 @@ message RuleType {
+                // -- minder.frizbee.actions.replace_tags_with_sha - finds any github actions within a workflow
+                //                                                   file and replaces the tag with the SHA using
+                //                                                   the frizbee library
+                string method = 3;
```
any better suggestions than method?

jhrozek replied:
1. empty string defaults to `minder.content`, in other words defaults to the old method, the only one that was available previously. This was probably not super visible in the large patch I submitted initially, sorry about that, but it's this helper function: 
```go
func getMethod(prCfg *pb.RuleType_Definition_Remediate_PullRequestRemediation) string {
	if prCfg.Method == "" {
		return minderContentModification
	}

	return prCfg.Method
}
```
2. this is "after I had ingested the original git repo, what do I do to produce a remediation"? With this PR, I've added calling frizbee to what we had earlier, which was basically evaluate a text template and replace a file with the expanded template contents.

jhrozek commented on proto/minder/v1/minder.proto:
```diff
@@ -1282,17 +1282,22 @@ message RuleType {
+                // the method to use to create the PR. For now, these are supported:
+                // -- minder.content - ensures that the content of the file is exactly as specified
+                //                     refer to the Content message for more details
+                // -- minder.frizbee.actions.replace_tags_with_sha - finds any github actions within a workflow
```
I wonder if frizbee should be in the name or just `minder.actions.replace_tags_with_sha` to not expose that we use frizbee under the hood as an implementation detail.

jhrozek replied:
done, marking as resolved

evankanderson reviewed (commented):
A few comments, I didn't get to all of `lib.go`.

jhrozek commented on internal/engine/actions/remediate/pull_request/lib.go:
yes, it's an internal library for for the pull_request module. I don't have a strong opinion on the naming.

jhrozek replied:
renamed, marking as done,

jhrozek commented on internal/engine/actions/remediate/pull_request/lib.go:
```diff
@@ -0,0 +1,352 @@
+	}
+
+	// #nosec G401 - we're not using sha1 for crypto, only to quickly compare contents
+	return fmt.Sprintf("%x", sha1.Sum([]byte(combinedContents))), nil
```
Is the slowness worth dragging in another dependency?

JAORMX replied:
I think that can be addressed in a separate PR.

jhrozek replied:
I have good experience with murmur hash from a previous project (in C), but I'd prefer not to drag another dependency for this. There's other codepaths that are way more of a bottleneck than the hash function.

jhrozek commented on internal/engine/actions/remediate/pull_request/lib.go:
```diff
@@ -0,0 +1,352 @@
+)
+
+type fsChanges interface {
+	sha1sum() (string, error)
```
will do, thanks

jhrozek commented on internal/engine/actions/remediate/pull_request/lib.go:
```diff
@@ -0,0 +1,352 @@
+		return fmt.Errorf("cannot parse mode: %w", err)
+	}
+
+	f, err := fs.OpenFile(fe.Path, os.O_RDWR|os.O_CREATE|os.O_TRUNC, os.FileMode(perms))
```
yes, that's how the old pull request remediation worked, too. The only thing it could do was to say "file X has contents Y", with the only extensibility being that the contents were a Go template. So this code just retains the old functionality.

Apart from this old functionality and the frizbee actions code, I'd like to also add merging (mergo probably?) and eventually retire this old file-clobbering functionality.

jhrozek commented on internal/engine/actions/remediate/pull_request/lib.go:
```diff
@@ -0,0 +1,352 @@
+type fsEntry struct {
+	contentTemplate *template.Template
+
+	Path    string `json:"path"`
+	Content string `json:"content"`
+	Mode    string `json:"mode"`
```
The JSON marshalling is only used in the dry run functionality which just dumps a JSON representation of the structure. The template member is also lower-cased so it won't be serialised.
(You are right that we don't have any tests whatsoever for the dry run and we should, I'll add at least a basic test)

jhrozek commented on internal/engine/actions/remediate/pull_request/lib.go:
```diff
@@ -0,0 +1,352 @@
+}
+
+func (fcs *fsChangeSet) writeEntries() error {
+	for i := range fcs.entries {
+		entry := fcs.entries[i]
```
I'm just neurotic about iterating over a slice of pointers in Go :-)
will change.

jhrozek commented on internal/engine/actions/remediate/pull_request/lib.go:
```diff
@@ -0,0 +1,352 @@
+			// just making sure we call sha1() after expandContents()
+			return "", fmt.Errorf("content (index %d) is empty", i)
+		}
+		combinedContents += fcs.entries[i].Path + fcs.entries[i].Content
```
I like this idea.

jhrozek commented on internal/engine/actions/remediate/pull_request/lib.go:
```diff
@@ -0,0 +1,352 @@
+
+	var combinedContents string
+
+	for i := range fcs.entries {
+		if len(fcs.entries[i].Content) == 0 {
+			// just making sure we call sha1() after expandContents()
+			return "", fmt.Errorf("content (index %d) is empty", i)
+		}
+		combinedContents += fcs.entries[i].Path + fcs.entries[i].Content
```
will take into account

jhrozek commented on internal/engine/actions/remediate/pull_request/lib.go:
```diff
@@ -0,0 +1,352 @@
+	if err != nil {
+		return fmt.Errorf("cannot marshal entries: %w", err)
+	}
+	fmt.Fprintf(out, "%s\n", b)
```
will take into account

jhrozek replied:
applied manually since the diff became outdated in the meantime

evankanderson commented on internal/engine/actions/remediate/pull_request/lib.go:
```diff
@@ -0,0 +1,352 @@
+		mode := ghModeNonExecFile
+		if cnt.Mode != nil {
+			mode = *cnt.Mode
+		}
```
Want to explicitly strip off the leading numbers if the length is >4 and the last 4 are an octal mode?  e.g.

```suggestion
		}
		// We used to use Git modes, which have a bunch of high-order bits.  Strip them out.
		if len(mode) > 4 && mode[len(mode)-4] == '0' {
			mode = mode[len(mode)-4:]
		}
```

jhrozek replied:
since we reverted to using git modes again, this comment is obsolete

jhrozek commented on proto/minder/v1/minder.proto:
```diff
@@ -1282,17 +1282,22 @@ message RuleType {
-                    // the GIT mode of the file. Not UNIX mode! String because the GH API also uses strings
-                    // the usual modes are: 100644 for regular files, 100755 for executable files and
-                    // 040000 for submodules (which we don't use but now you know the meaning of the 1 in 100644)
+                    // The UNIX file permissions to set on the file
```
actually, I was stupid and git doesn't support the full set of UNIX permissions, but a subset of them plus some of its own (see e.g. https://github.com/go-git/go-git/blob/32e0172851c35ae2fac495069c923330040903d2/plumbing/filemode/filemode.go#L10)

So I reverted to using the git mode back and I actually used go-git's utility functions to convert the git modes into unix mode with error handling.

Since we're squashing PR patches, I'll be pushing the changes as individual patches for easier review.

jhrozek commented on pkg/providers/v1/providers.go:
```diff
@@ -107,6 +107,7 @@ type GitHub interface {
 	CreatePullRequest(ctx context.Context, owner, repo, title, body, head, base string) (*github.PullRequest, error)
 	ListPullRequests(ctx context.Context, owner, repo string, opt *github.PullRequestListOptions) ([]*github.PullRequest, error)
 	CreateComment(ctx context.Context, owner, repo string, number int, comment string) error
+	GetClient() *github.Client
```
done, marking as resolved.

jhrozek commented:
sorry about the conference-induced delay in getting to your comments @evankanderson and @JAORMX. I pushed a new version that addresses most of your comments and will address the rest tomorrow!

jhrozek commented on internal/engine/actions/remediate/pull_request/lib.go:
```diff
@@ -0,0 +1,352 @@
+	fsChanges
+	createFsModEntries(ctx context.Context, params interfaces.ActionsParams) error
+	modifyFs(ctx context.Context, params interfaces.ActionsParams) ([]*fsEntry, error)
+}
```
it was just easier for me to build the mental model between the two sets of structures (fsChangeset that is embedded in the actual implementations and the implementations themselves). No other reason. I merged the interfaces.

jhrozek commented on internal/engine/actions/remediate/pull_request/lib.go:
```diff
@@ -0,0 +1,352 @@
+func (ca *contentModification) prConfigToEntries() ([]*fsEntry, error) {
+	entries := make([]*fsEntry, len(ca.prCfg.Contents))
+	for i := range ca.prCfg.Contents {
+		cnt := ca.prCfg.Contents[i]
```
applied manually since the diff became outdated in the meantime

jhrozek commented on internal/engine/actions/remediate/pull_request/lib.go:
```diff
@@ -0,0 +1,352 @@
+		contentTemplate, err := util.ParseNewTextTemplate(&cnt.Content, fmt.Sprintf("Content[%d]", i))
+		if err != nil {
+			return nil, fmt.Errorf("cannot parse content template (index %d): %w", i, err)
+		}
```
this is what I started with actually, but I didn't like that the entries are created in the constructor, then not used until the modification is (or is not) called. But you have a point with throwing the error soon.

jhrozek commented on internal/engine/actions/remediate/pull_request/lib.go:
```diff
@@ -0,0 +1,352 @@
+		}
+
+		mode := ghModeNonExecFile
+		if cnt.Mode != nil {
```
good idea, applied

jhrozek commented on internal/engine/actions/remediate/pull_request/lib.go:
```diff
@@ -0,0 +1,352 @@
+		"Profile": params.GetRule().Def.AsMap(),
+	}
+	for i := range entries {
+		entry := entries[i]
```
applied

jhrozek commented on internal/engine/actions/remediate/pull_request/lib.go:
```diff
@@ -0,0 +1,352 @@
+		return nil, fmt.Errorf("cannot write entries: %w", err)
+	}
+	return ftr.entries, nil
+}
```
yes, I would prefer slightly more duplication now and cleaner interfaces.

jhrozek replied:
embarrassingly you were right, I was calling createFSModEntries from the calling site and ALSO here..which I only found when unit tests required me to mock upstream calls twice..

jhrozek commented:
@JAORMX @evankanderson I hope all comments are addressed and with the new frizbee release which accepts an interface I was also able to write a unit tests. PTAL..

evankanderson commented on internal/engine/actions/remediate/pull_request/internal.go:
```diff
@@ -0,0 +1,343 @@
+
+const (
+	minderContentModification = "minder.content"
+	minderFrizbeeTagResolve   = "minder.actions.replace_tags_with_sha"
```
(Different from the docs in `minder.proto`?)

jhrozek commented:
Excluding actions is tracked here: https://github.com/stacklok/minder/issues/1910
